src/call_reply.c

```diff
@@ -0,0 +1,487 @@
+    callReplyParse(rep);
+    if (rep->type != type) return C_ERR;
+    if (idx >= rep->len) return C_ERR;
+    if (key) *key = callReplyGetCollectionElement(rep, idx * 2, 2);
+    if (val) *val = callReplyGetCollectionElement(rep, idx * 2 + 1, 2);
+    return C_OK;
+}
+
+/**
+ * Retrieve map reply key and value at a given index, applicabale for:
+ * * REDISMODULE_REPLY_MAP
+ *
+ * The key and val are both output params which can be NULL(in this case they are not populated).
+ * Return C_OK on success and C_ERR if reply type is wrong or if the idx is out of range.
+ *
+ * The returned values are only borrowed and their lifetime is
+ * as long as the given CallReply. In addition there is no need
+ * to manually free the returned CallReplies, it will be freed when
+ * the root CallReplied will be freed.
+ */
+int callReplyGetMapElement(CallReply* rep, size_t idx, CallReply** key, CallReply** val) {
+    return callReplyGetMapElementInternal(rep, idx, key, val, REDISMODULE_REPLY_MAP);
+}
+
+/**
+ * Return reply attribute if exists or NULL, applicabale for all replies.
+ *
+ * The returned value is only borrowed and its lifetime is
+ * as long as the given CallReply. In addition there is no need
+ * to manually free the returned CallReply, it will be freed when
+ * the root CallReplied will be freed.
+ */
+CallReply* callReplyGetAttribute(CallReply* rep) {
+    return rep->attribute;
+}
+
+/**
+ * Retrieve attribute reply key and value at a given index, applicabale for:
+ * * REDISMODULE_REPLY_ATTRIBUTE
+ *
+ * The key and val are both output params which can be NULL(in this case they are not populated).
+ * Return C_OK on success and C_ERR if reply type is wrong or if the idx is out of range.
+ *
+ * The returned values are only borrowed and their lifetime is
+ * as long as the given CallReply. In addition there is no need
+ * to manually free the returned CallReplies, it will be freed when
+ * the root CallReplied will be freed.
+ */
+int callReplyGetAttributeElement(CallReply* rep, size_t idx, CallReply** key, CallReply** val) {
+    return callReplyGetMapElementInternal(rep, idx, key, val, REDISMODULE_REPLY_MAP);
+}
+
+/**
+ * Return big number reply value, applicabale for:
+ * * REDISMODULE_REPLY_BIG_NUMBER
+ *
+ * The returned value is only borrowed and its lifetime is
+ * as long as the given CallReply.
+ * The returned value is promised to be a big number as describe
+ * on RESP3 spacifications.
+ * The returned value is not NULL terminated and its mandatory to
+ * give the len argument
+ */
+const char* callReplyGetBigNumber(CallReply* rep, size_t* len) {
+    callReplyParse(rep);
+    if (rep->type != REDISMODULE_REPLY_BIG_NUMBER) return NULL;
+    *len = rep->len;
+    return rep->val.str;
+}
+
+/**
+ * Return verbatim string reply value, applicabale for:
+ * * REDISMODULE_REPLY_VERBATIM_STRING
+ *
+ * An optional output argument can be given to get verbatim reply
+ * format, NULL can be give to ignore it.
```

Comment:
came here to change `give` to `given`, but ended up changing the wording. you can choose otherwise.
```suggestion
 * format, or NULL if not needed.
```